### DISCUSSION TITLE
Is there an official way to create multiple service providers from one web host?

### DISCUSSION THREAD

**Discussion by the-avid-engineer** [dotnet/aspnetcore#27420] [Q&A]
Currently you can use `app.MapWhen(...)` to have a per-route middleware stack. Unfortunately, these middlewares all share the same service provider.

I found this issue on the old repository: https://github.com/aspnet/Mvc/issues/5326

@rynowak responded in part
>If the answer is one site then you'll hit a point where you need two different service providers. We've done some experimenting with this but don't have a published solution yet. The idea with this is to have a 'dispatcher' app that receives requests, and then forwards them to a 'tenant' app which processes. Because each tenant has it's own services, it's totally isolated. The fact that the controllers happen to be in the same assembly is just a details.

That was four years ago.. I'm curious if that experimenting has gotten anywhere? I'm not sure where to look for this kind of thing.

Someone made a workaround, but I'm not particularly keen on the idea because it creates additional web hosts. https://www.strathweb.com/2017/04/running-multiple-independent-asp-net-core-pipelines-side-by-side-in-the-same-application/

**Comment by davidfowl** [comment]
cc @sebastienros

**Reply by Tratcher** [reply]
Sharing a container is OK if you only need scoped services.

And no, there's no built in support yet for multiple containers without creating separate hosts. Orchard Core is the only place I know that's built that functionality.
